# Bài tập thực hành 1

In [7]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')


In [2]:
df= pd.read_csv("pd_speech_features.csv")
df.head()
df['class'].value_counts()

class
1    564
0    192
Name: count, dtype: int64

a) Giảm số chiều xuống còn 200 ứng với thành phần chính của dữ liệu bằng PCA nhằm phục vụ cho việc phân loại dữ liệu (theo class), với phần dữ liệu Train gồm 500 bản ghi và dữ liệu kiểm tra là phần còn lại. Dùng các công cụ đo độ chính xác để kiểm tra đánh giá mô hình.

In [3]:
y = df['class']
X = df.drop(["id", "class"], axis =1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=len(df)-500, shuffle=False)

In [4]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [5]:
# Áp dụng PCA
pca = PCA(n_components=200)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

In [6]:
# Phân loại dữ liệu
classifier = LogisticRegression()
classifier.fit(X_train_pca, y_train)
y_pred = classifier.predict(X_test_pca)

In [7]:
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:")
print(conf_matrix)

Accuracy: 0.7890625
Confusion Matrix:
[[ 54  27]
 [ 27 148]]


b) Hãy sử dụng phương phap PCA để đưa dữ liệu về số chiều nhỏ nhất sao cho lượng thông tin được giữ lại ít nhất đạt 80%.

- i. Cho biết dữ liệu mới còn bao nhiêu chiều.

In [8]:
yb = df['class']
Xb = df.drop(['id', 'class'], axis=1)

In [9]:
scaler = StandardScaler()
Xb = scaler.fit_transform(Xb)

pca = PCA(n_components=0.8)
Xb_pca = pca.fit_transform(Xb)

In [10]:
new_dimensions = Xb_pca.shape[1]
print("New Dimensions:", new_dimensions)

New Dimensions: 61


- ii. Thực hiện lại bài toán phân loại với tỷ lệ như trên, ứng với dữ liệu mới và so sánh độ chính xác.

In [11]:
Xb_train, Xb_test, yb_train, yb_test = train_test_split(Xb_pca, yb, test_size=len(X) - 500, shuffle=False)

In [12]:
# Phân loại dữ liệu
classifier = LogisticRegression()
classifier.fit(Xb_train, yb_train)
yb_pred = classifier.predict(Xb_test)

In [13]:
accuracy_b = accuracy_score(yb_test, yb_pred)
conf_matrix_b = confusion_matrix(yb_test, yb_pred)

print("Accuracy:", accuracy_b)
print("Confusion Matrix:")
print(conf_matrix_b)

Accuracy: 0.796875
Confusion Matrix:
[[ 52  29]
 [ 23 152]]


c) Sử dụng dữ liệu ban đầu, chia thành các tập Train- Test với tỷ lệ 4:2, sau đó áp dụng phương pháp Naïve Bayes phù hợp và phương pháp Hồi quy Logistic để thực hiện bài toán phân loại (theo class).
Tiếp theo lại thực nghiệm các mô hình nói trên với dữ liệu đã giảm chiều ở ý b), cùng tỷ lệ chia như trên. Hãy đánh giá xem 2 mô hình này, mô hình nào có độ chính xác thay đổi nhiều hơn.

In [14]:
yc = df['class']
Xc = df.drop(['id', 'class'], axis=1)

In [15]:
# Divide the data set into train and test sets with a ratio of 4:2
Xc_train, Xc_test, yc_train, yc_test = train_test_split(Xc, yc, test_size=0.4, random_state=42)

In [16]:
# Naive Bayes on original data
nb_classifier = GaussianNB()
nb_classifier.fit(Xc_train, yc_train)

GaussianNB()

In [17]:
nb_pred = nb_classifier.predict(Xc_test)
nb_accuracy_original = accuracy_score(yc_test, nb_pred)

In [18]:
# Logistic Regression on original data
lr_classifier = LogisticRegression()
lr_classifier.fit(Xc_train, yc_train)

LogisticRegression()

In [19]:
lr_pred = lr_classifier.predict(Xc_test)
lr_accuracy_original = accuracy_score(yc_test, lr_pred)

In [20]:
# Divide the reduced-dimensional data set into train and test sets with a ratio of 4:2

X_train_pca, X_test_pca, y_train, y_test = train_test_split(Xb_pca, yb, test_size=0.4, random_state=42)

In [21]:
# Naive Bayes on dimensionally reduced data
nb_classifier_pca = GaussianNB()
nb_classifier_pca.fit(X_train_pca, y_train)

GaussianNB()

In [22]:
nb_pred_pca = nb_classifier_pca.predict(X_test_pca)
nb_accuracy_pca = accuracy_score(y_test, nb_pred_pca)

In [23]:
# Logistic Regression on reduced-dimensional data
lr_classifier_pca = LogisticRegression()
lr_classifier_pca.fit(X_train_pca, y_train)

LogisticRegression()

In [24]:
lr_pred_pca = lr_classifier_pca.predict(X_test_pca)
lr_accuracy_pca = accuracy_score(y_test, lr_pred_pca)

In [25]:
#Evaluate the accuracy of the models

print("Naive Bayes accuracy on original data:", nb_accuracy_original)
print("Accuracy of Logistic Regression on original data:", lr_accuracy_original)

print("Naive Bayes accuracy after dimensionality reduction:", nb_accuracy_pca)
print("Accuracy of Logistic Regression after dimensionality reduction:", lr_accuracy_pca)

Naive Bayes accuracy on original data: 0.6897689768976898
Accuracy of Logistic Regression on original data: 0.7161716171617162
Naive Bayes accuracy after dimensionality reduction: 0.7161716171617162
Accuracy of Logistic Regression after dimensionality reduction: 0.801980198019802


In [26]:
# model accuracy changes
print("value changes of Navie Bayes:", abs(nb_accuracy_original - nb_accuracy_pca))
print("value changes of Logistic Regression:", abs(lr_accuracy_original - lr_accuracy_pca))

value changes of Navie Bayes: 0.02640264026402639
value changes ofLogistic Regression: 0.08580858085808585


----> model Logistic có sự thay đổi nhiều hơn

# Bài tập thực hành 2

1) Đọc dữ liệu ảnh, lấy tập dữ liệu 5000 ảnh bất kỳ; giảm số chiều dữ liệu xuống còn 100 chiều (từ 28x28 = 784 chiều ban đầu).

In [8]:
import os
import numpy as np
from sklearn.linear_model import LogisticRegression
import time

In [9]:
data_path = 'C:/Users/Admin/Desktop/ML/week4'

# train path
train_images_path = os.path.join(data_path, 'train-images-idx3-ubyte.gz')
train_labels_path = os.path.join(data_path, 'train-labels-idx1-ubyte.gz')

# test path
test_images_path = os.path.join(data_path, 't10k-images-idx3-ubyte.gz')
test_labels_path = os.path.join(data_path, 't10k-labels-idx1-ubyte.gz')

In [10]:
def get_mnist_data(images_path, labels_path, num_images, shuffle=False, _is=True, image_size=28):
    """
    This shuffle param is active when .gz is downloaded at:
    - 'http://yann.lecun.com/exdb/mnist/'
    - This function return random num_images in 60000 or 10000
    """
    # read data
    import gzip # to decompress gz (zip) file
    
    # open file training to read training data
    f_images = gzip.open(images_path,'r')
    
    # skip 16 first bytes because these are not data, only header infor
    f_images.read(16)
    
    # general: read num_images data samples if this parameter is set;
    # if not, read all (60000 training or 10000 test)
    real_num = num_images if not shuffle else (60000 if _is else 10000)
    
    # read all data to buf_images (28x28xreal_num)
    buf_images = f_images.read(image_size * image_size * real_num)
    
    # images
    images = np.frombuffer(buf_images, dtype=np.uint8).astype(np.float32)
    images = images.reshape(real_num, image_size, image_size,)
    
    # Read labels
    f_labels = gzip.open(labels_path,'r')
    f_labels.read(8)
    
    labels = np.zeros((real_num)).astype(np.int64)
    
    # rearrange to correspond the images and labels
    for i in range(0, real_num):
        buf_labels = f_labels.read(1)
        labels[i] = np.frombuffer(buf_labels, dtype=np.uint8).astype(np.int64)

    # shuffle to get random images data
    if shuffle is True:
        rand_id = np.random.randint(real_num, size=num_images)
        
        images = images[rand_id, :]
        labels = labels[rand_id,]
        
    # change images data to type of vector 28x28 dimentional
    images = images.reshape(num_images, image_size * image_size)
    return images, labels

In [11]:
data, label = get_mnist_data(train_images_path, train_labels_path, 5000, shuffle=True)

In [12]:
scaler = StandardScaler()
X = scaler.fit_transform(data)

In [13]:
pca = PCA(n_components=100)
X_pca = pca.fit_transform(X)

In [14]:
print("New dimension after data dimension reduction:", X_pca.shape[1])

New dimension after data dimension reduction: 100


2) Áp dụng phương pháp phân loại nhiều lớp Multinomial Logistic Regression (tham khảo bài trước) để phân loại để phân loại tập dữ liệu đã chọn trong ý 1), tỷ lệ train:validation là 0.7:0.3.
Hãy so sánh kết quả (độ chính xác và thời gian) chạy mô hình phân loại trong hai trường hợp:

- a) Dữ liệu nguyên bản (giữ nguyên 28x28 = 784 chiều)

In [15]:
Xr_train, Xr_test, yr_train, yr_test = train_test_split(X, label, test_size=0.3, random_state=42)

In [16]:
raw_data_start_time = time.time()

mlr = LogisticRegression(multi_class='multinomial', max_iter=5000, solver='sag')
mlr.fit(Xr_train, yr_train)

raw_data_end_time = time.time()

In [17]:
yr_pred = mlr.predict(Xr_test)
raw_data_accuracy = accuracy_score(yr_test, yr_pred)
raw_data_elapse_time = raw_data_end_time - raw_data_start_time

In [18]:
print("time elapse", raw_data_elapse_time)
print("accuracy", raw_data_accuracy)

time elapse 294.1235382556915
accuracy 0.882


- b) Dữ liệu đã qua giảm chiều (100 chiều). Chú ý cần thực hiện giảm chiều với tập ảnh trước, sau đó mới chia thành tập train và validation.

In [23]:
X_pca_train, X_pca_test, y_pca_train, y_pca_test = train_test_split(X_pca, label, test_size = 0.3, random_state=42)

In [24]:
pca_start_time = time.time()

mlr_pca = LogisticRegression(multi_class='multinomial', max_iter=5000, solver='sag')
mlr_pca.fit(X_pca_train, y_pca_train)

pca_end_time = time.time()

In [25]:
y_pca_pred = mlr_pca.predict(X_pca_test)
pca_accuracy = accuracy_score(y_pca_test, y_pca_pred)
pca_elapse_time = pca_end_time - pca_start_time

In [26]:
print("time elapse", pca_elapse_time)
print("accuracy", pca_accuracy)

time elapse 36.23029637336731
accuracy 0.8646666666666667


so sánh kết quả (độ chính xác và thời gian) chạy mô hình

- Độ chính xác:
    + dữ liệu nguyên bản = 0.882
    + dữ liệu đã giảm chiều = 0.8646666666666667 <br>
    -> dữ liệu nguyên bản nhỏ hơn dữ liệu đã giảm chiều (chỉ có 0.0173333333333333)

- Thời gian chạy:
    + dữ liệu nguyên bản = 294.1235382556915 s
    + dữ liệu đã giảm chiều = 36.23029637336731 s<br>
    -> dữ liệu nguyên bản lớn hơn dữ liệu đã giảm chiều (tận 257.89324188232419 giây)